File: nuxt_side/components/models/safe_sfen.js
import { GX } from "@/components/models/gs.js"

export const SafeSfen = {
  encode(sfen) {
    return GX.urlsafe_encode64(sfen)
  },

  // Rails側と合わせていれているがまだどこでも使っていない
  decode(bin) {
    return GX.urlsafe_decode64(bin)
  },
}

File: app/models/safe_sfen.rb
module SafeSfen
  extend self

  def encode(sfen)
    Base64.urlsafe_encode64(sfen, padding: false)
  end

  # KI2の場合を考慮してUTF-8化する
  # UTF-8 化しないと Encoding::CompatibilityError (incompatible character encodings: ASCII-8BIT and UTF-8) になる
  def decode(bin)
    Base64.urlsafe_decode64(bin).toutf8
  end
end
